torch/custom_class.h

```diff
@@ -221,6 +221,82 @@ class class_ {
     return *this;
   }
 
+  // This is a temporary hack to change the format of ConvPackedParams without
+  // breaking BC. We modify the default pickle protocol by defining a boxed
+  // version of __set_state__, which knows how to parse historical formats.
+  // TODO(before land): create an issue to delete this in a month and link here.
+  template <typename GetStateFn, typename SetStateFn>
+  class_& def_pickle_unboxed_DO_NOT_USE(
+      GetStateFn&& get_state, SetStateFn&& set_state) {
+    // Note: most of this code is copy-pasta from def_pickle
+    // TODO(before land): if this approach gets consensus, potentially reuse
+    //   more things.
+    static_assert(
+        c10::guts::is_stateless_lambda<std::decay_t<GetStateFn>>::value &&
+            c10::guts::is_stateless_lambda<std::decay_t<SetStateFn>>::value,
+        "def_pickle() currently only supports lambdas as "
+        "__getstate__ and __setstate__ arguments.");
+    def("__getstate__", std::forward<GetStateFn>(get_state));
+
+    // __setstate__ needs to be registered with some custom handling:
+    // We need to wrap the invocation of of the user-provided function
+    // such that we take the return value (i.e. c10::intrusive_ptr<CurrClass>)
+    // and assign it to the `capsule` attribute.
+    using SetStateTraits =
+        c10::guts::infer_function_traits_t<std::decay_t<SetStateFn>>;
+    using SetStateArg = typename c10::guts::typelist::head_t<
+        typename SetStateTraits::parameter_types>;
+    auto setstate_wrapper = [set_state = std::move(set_state)](
+                                c10::tagged_capsule<CurClass> self,
+                                SetStateArg&& arg) {
+      c10::intrusive_ptr<CurClass> classObj =
+          at::guts::invoke(set_state, std::forward<SetStateArg>(arg));
+      auto object = self.ivalue.toObject();
+      object->setSlot(0, c10::IValue::make_capsule(classObj));
+    };
+    defineMethod(
```

Comment:
hm, how does it even work? I don't think over schema inference can handle IValue tbh. Can you print out the schema of what gets registered (i.e. `setstate_schema ` below)?

I'd expect this to be `_def_unboxed` call (actually the name is wrong - it should be `_def_boxed`). And as the schema you can construct the schema explicitly by taking the type of the argument from `getstate`.

The idea is that you want the schema for `__setsttate__` to be correct to make it invocable from Torchscript (not used right now but shall we do it ever). But the implementation can be "generic" and accept just IValue (when invoked by pickler).

Maybe I miss something and our schema actually support IValue-valued args. In which case your implementation might be ok